### PR TITLE
perf: Performance improvements to makeFullMesh function in MapLayer  #1662

### PR DESCRIPTION
**Pull request by nilscb**
Optimizations done:
   - Ability to specify typed arrays as input (Float32Array).
   - Pre-allocate and use typed arrays in internal calculation of mesh.
   - Use same vertices for lines and triangles.
   - Only calculate lines and normals if specified.
   - Remove webworker.
   - Never copy input or change input data arrays.
   - Specify normals as signed bytes (int8) to save space.
   - General cleanup.

### CODE REVIEW

**Comment by t0oF-azpn** [comment]
I saw that you did some perf measures and that you removed the webworker.
What size of grid did you check (and what where the perfs) ?

**Comment by t0oF-azpn** [comment]
BTW, add a description to summarize the optimizations you apply in this PR

**Comment by nilscb** [comment]
In the story "BigMap" in webiz which is a map with 1452x1141 nodes the loading time went down from approximately 0.75 sec to 0.2 sec, I will investigate a bit the memory usage.

**Comment by nilscb** [comment]
I added a description of what has been done above.

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.stories.tsx [hunk 1]
```diff
@@ -871,6 +873,91 @@ NodeCenteredPropMapWithArrayInput.parameters = {
...
+    const f = makeGaussian(amplitude, X0, Y0, stdX, stdY);
+
+    const data = new Float32Array(n * n).map((val, index) => {
+        const x = (index % n) - n / 2;
+        const y = Math.floor(index / n) - n / 2;
+        return f(x, y); // keep + 0.3 * Math.random();
+    });
+
+    const t1 = performance.now();
+    console.log(`makeData took ${(t1 - t0) * 0.001}  seconds.`);
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/map/mapLayer.stories.tsx:901]: Use `console.debug` for debugging info

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts [hunk 2]
```diff
@@ -131,15 +114,17 @@ export default class privateMapLayer extends Layer<privateMapLayerProps> {
             vs: vsShader,
             fs: fsShader,
             geometry: new Geometry({
-                drawMode: this.props.mesh.drawMode,
+                drawMode: 4, // triangles
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts:117]: Is there a constant defined for this?

**Review by hkfb — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.stories.tsx [hunk 3]
```diff
@@ -871,6 +873,91 @@ NodeCenteredPropMapWithArrayInput.parameters = {
...
+    return data;
+}
+
+//-- NodeCenteredPropMap  with native javascript arrays as input --
+// Mesh will be n x n nodes.
+const n = 300;
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/map/mapLayer.stories.tsx:907]: Would be nice to expose these variables as arguments to the story. There are some examples in some other stories. I can give some pointers.

**Review by t0oF-azpn — APPROVED** [review]

File: typescript/packages/subsurface-viewer/src/layers/map/fragment.fs.glsl.ts [hunk 4]
```diff
@@ -34,21 +34,19 @@ uniform bool isColorMapClampColorTransparent;
...
+   // These are sent as Int8
+   normal[0] /= 127.0;
+   normal[1] /= 127.0;
+   normal[2] /= 127.0;
 
    if (!smoothShading) {
       normal = normalize(cross(dFdx(position_commonspace.xyz), dFdy(position_commonspace.xyz)));
-   } 
-
-   // // Discard transparent pixels. KEEP
```
[the remark this answers is not in the document]
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/map/fragment.fs.glsl.ts:46]: This code is not needed anymore hence I removed it along with its comments.

File: typescript/packages/subsurface-viewer/src/layers/map/fragment_lines.glsl.ts [hunk 5]
```diff
@@ -7,13 +7,6 @@ precision highp float;
 out vec4 fragColor;
 
 void main(void) {
-  
-  // Picking pass.
```
[the remark this answers is not in the document]
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/map/fragment_lines.glsl.ts:11]: Unnecessary comment as the next line is self explanatory.

File: typescript/packages/subsurface-viewer/src/layers/map/utils.ts [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+    const i3 = h * nx + (w + 1);
+    const i4 = (h - 1) * nx + w;
+
+    const i0_act =                 isDefined(meshData[i0]); // eslint-disable-line
+    const i1_act = (w - 1) >= 0 && isDefined(meshData[i1]); // eslint-disable-line
+    const i2_act = (h + 1) < ny && isDefined(meshData[i2]); // eslint-disable-line
+    const i3_act = (w + 1) < nx && isDefined(meshData[i3]); // eslint-disable-line
+    const i4_act = (h - 1) >= 0 && isDefined(meshData[i4]); // eslint-disable-line
+
+    const noNormal = [0, 0, 1]; // signals a normal could not be calculated.
```
[the remark this answers is not in the document]
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/map/utils.ts:70]: Good point.Changed it to [0,0,0]

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts [hunk 7]
```diff
@@ -54,15 +54,16 @@ export type Params = {
...
 async function load_mesh_and_properties(
-    meshData: string | number[],
-    propertiesData: string | number[],
-    ZIncreasingDownwards: boolean
+    meshData: string | number[] | Float32Array,
+    propertiesData: string | number[] | Float32Array
 ) {
     // Keep
-    //const t0 = performance.now();
+    const t0 = performance.now();
```
[the remark this answers is not in the document]
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts:66]: Commented out.

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts [hunk 8]
```diff
@@ -54,15 +54,16 @@ export type Params = {
     isMesh: boolean;
     frame: Frame;
     smoothShading: boolean;
+    gridLines: boolean;
+    ZIncreasingDownwards: boolean;
 };
 
 async function load_mesh_and_properties(
```
[the remark this answers is not in the document]
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts:61]: I guess this is for the user if he wants lines or not. The user must assume the most efficient way are used internally.

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts [hunk 9]
```diff
@@ -163,15 +170,9 @@ async function load_mesh_and_properties(
...
-    if (!ZIncreasingDownwards) {
-        for (let i = 0; i < mesh.length; i++) {
-            mesh[i] *= -1;
-        }
-    }
-
-    //const t1 = performance.now();
     // Keep this.
-    //console.log(`Task loading took ${(t1 - t0) * 0.001}  seconds.`);
+    const t1 = performance.now();
```
[the remark this answers is not in the document]
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts:174]: Done

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts [hunk 10]
```diff
@@ -150,8 +135,14 @@ export default class privateMapLayer extends Layer<privateMapLayerProps> {
...
-            geometry: new Geometry(this.props.meshLines),
-            modules: [project, picking],
+            geometry: new Geometry({
+                drawMode: 1, // lines
+                attributes: {
+                    positions: { value: this.props.positions, size: 3 },
+                },
+                indices: { value: this.props.lineIndices, size: 1 },
+            }),
+            modules: [project],
```
[the remark this answers is not in the document]
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts:145]: Not for lines!

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.stories.tsx (hunk 3 above)
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/map/mapLayer.stories.tsx]: Done

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts (hunk 2 above)
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: Used standard constants

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (t0oF-azpn) [reply typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts:61]: I meant some comment in the code.
It's a pain to have to read the code of a function during a code review to understand what it's doing...

**Review by t0oF-azpn — APPROVED** [review]

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.stories.tsx [hunk 11]
```diff
@@ -871,6 +873,111 @@ NodeCenteredPropMapWithArrayInput.parameters = {
...
+                    id: "view_1",
+                    show3D: true,
+                },
+            ],
+        },
+    };
+    return <SubsurfaceViewer {...subsurfaceViewerArgs} />;
+};
+
+export const TypedArrayInput: ComponentStory<typeof SubsurfaceViewer> = (
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/map/mapLayer.stories.tsx:955]: `ComponentStory` is deprecated in Storybook > v6

**Review by hkfb — APPROVED** [review]

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.stories.tsx (hunk 11 above)
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/map/mapLayer.stories.tsx]: Ok I changed it to "StoryFn"

**Pull request merged**